### DISCUSSION TITLE
Testing python cli tool using typer -- Trying to change home dir

### DISCUSSION THREAD

**Discussion by ajoino** [tox-dev/tox#2254] [Q&A] [answered]
Hi.

I'm developing a CLI tool using typer, and during installation with pip the tool finds and creates a directory in `~/.config` on Linux, achieved through typer's `get_app_dir` function. This directory will then contain information necessary to run the tool.

This has worked fine thus far, but now I want to run tests with tox and I run into the issue that the tests will also use my user's directory, which means the tests cannot test installations in a clean directory, and the tests will also be contaminated with my previous data generated and stored in the config directory.

I've searched but I cannot find a solution to this problem, my last try was to edit the `$HOME` environment variable but that doesn't play nicely with typer. So I am wondering if it is possible, using tox or pytest, to basically 'mock' the file system when running tests to make `$HOME`/`~` point to another place, and preferably also make and remove the directory during startup and teardown.

Kind regards
Jacob Nilsson

**Comment by jugmac00** [comment, marked as answer]
This cannot be solved by `tox` - `tox` just drives `pytest` or any other test runner, but does not test and also not create mocks itself.

I see a couple of options. If the path which points to `~/.config` is exposed, you could mock or inject it.

You possibly could mock `typer`'s `get_app_dir`, although this may cause side effects.

Finally, you could use `pyfakefs` which creates a fake file system for the tests only, see http://jmcgeheeiv.github.io/pyfakefs/master/usage.html

For `pytest` questions - and this is a `pytest` question - you could also visit their discord server:
https://discord.com/invite/pytest-dev

**Reply by ajoino** [reply]
Thank you for the fast and detailed reply!

I am not well versed in the Python testing world (though I'm trying to become) and it is sometimes unclear to me what tool should be used for certain tasks. In this case, it seemed like `tox` was the appropriate tool, but I know understand that it's not. I really appreciate you suggesting alternative solutions, I'll try them out
